brave/src/main/java/brave/propagation/B3SingleFormat.java

```diff
@@ -158,79 +171,98 @@ public static TraceContextOrSamplingFlags parseB3SingleFormat(CharSequence value
-    } else if (traceIdLength == 16) {
-      traceIdHigh = 0L;
-    } else {
-      Platform.get().log("Invalid input: expected a 16 or 32 lower hex trace ID", null);
-      return null;
-    }
+    // Instead of pos < endIndex, this uses pos <= endIndex to keep field processing consolidated.
+    // Otherwise, we'd have to process again when outside the loop.
+    for (int pos = beginIndex; pos <= endIndex; pos++) {
+      // treat EOF same as a hyphen for simplicity
+      boolean isEof = pos == endIndex;
+      char c = isEof ? '-' : value.charAt(pos);
 
-    traceId = tryParseHex("trace ID", value, pos, endIndex);
-    if (traceId == 0) return null;
-    pos += 17; // lower 64 bits of the trace ID and the hyphen
+      if (c == '-') {
+        if (currentField == Field.SAMPLED) {
+          // The last field could be sampled or parent ID. Revise assumption if longer than 1 char.
+          if (isEof && currentFieldLength > 1) {
+            currentField = Field.PARENT_SPAN_ID;
+          }
+        }
 
-    long spanId = tryParseHex("span ID", value, pos, endIndex);
-    if (spanId == 0) return null;
-    pos += 16; // spanid
+        if (!validateFieldLength(currentField, currentFieldLength)) {
+          return null;
+        }
 
-    int flags = 0;
-    long parentId = 0L;
-    if (hyphenCount > 1) { // traceid-spanid-
-      pos++; // consume the hyphen
-
-      if (hyphenCount == 3) { // we should parse sampled AND parent ID
-        flags = tryParseSampledFlags(value, endIndex, pos);
-        if (flags == 0) return null;
-        pos += 2; // consume the sampled flag and hyphen
-        parentId = tryParseParentId(value, endIndex, pos);
-        if (parentId == 0L) return null;
-      } else { // we should parse sampled OR parent ID
-        if (endIndex - pos <= 1) {
-          flags = tryParseSampledFlags(value, endIndex, pos);
-          if (flags == 0) return null;
+        if (currentField.required && buffer == 0L) {
+          Platform.get().log("Invalid input: read all zeroes {0}", currentField.name, null);
+          return null;
+        }
+
+        if (currentField.equals(Field.TRACE_ID)) {
+          traceId = buffer;
+
+          currentField = Field.SPAN_ID;
+        } else if (currentField.equals(Field.SPAN_ID)) {
+          spanId = buffer;
+
+          // The handle malformed cases like below, it is easier to assume the next field is sampled
+          // and revert if definitely not vs try to determine if it is definitely sampled.
+          // 'traceId-spanId--parentSpanId'
+          // 'traceId-spanId-'
+          currentField = Field.SAMPLED;
+        } else if (currentField.equals(Field.SAMPLED)) {
+          SamplingFlags samplingFlags = tryParseSamplingFlags(value.charAt(pos - 1));
+          if (samplingFlags == null) return null;
+          flags = samplingFlags.flags;
+
+          currentField = Field.PARENT_SPAN_ID;
+          currentFieldLength = 0;
+          continue;
```

Comment:
Reasoning this `continue` out is tough since it's buried among much code. IIUC, this just avoids a single `buffer = 0L` so maybe not worth special casing?

Reply:
this was fuzz indeed.. good eye!